### ISSUE TITLE
RDFa parser/writer

### ISSUE THREAD

**Issue by jeenbroekstra** [eclipse/rdf4j#512]
(Migrated from https://openrdf.atlassian.net/browse/SES-951)

See http://www.w3.org/TR/xhtml-rdfa-primer/ The current Sesame trunk contains an RDFa parser/writer. It would be good to see if it can be backported to 2.x and updated to the latest set of specs, so that Sesame/Rio have full RDFa support.

**Comment by barthanssens** [comment]
I've found this code https://github.com/ansell/openrdf-sesame/tree/3.0-archive/core/rio/rdfa by @jamesrdf.

Since Google etc is promoting semantic markup on websites (https://developers.google.com/search/docs/data-types/dataset, even though JSON-LD seems to be preferred), there is some renewed interest in RDFa (or maybe at least parsing RDFa Lite)

Would it make sense revive the code but use e.g. jsoup instead of an xslt-based approach, since "modern" websites nowadays aim to be HTML5 instead of XHTML ?

**Comment by barthanssens** [comment]
AttoParser might be an interesting (event-based) alternative https://github.com/attoparser/attoparser

**Comment by jeenbroekstra** [comment]
Yeah, the main issue with that old sesame 3 code is the fact that it relies on XSLT (also, secondary: the IP rights on that code, especially the stylesheet itself, are a tad murky, so we might be better off starting from scratch and using as inspiration only). 

@barthanssens I'm curious: do you have a use case where you might need RDFa support in RDF4J?

**Comment by tfrancart** [comment]
possible use-cases include website crawler + structured data extraction to feed knowledge graph. Or  : SHACL-based validation of structured markup.

**Comment by barthanssens** [comment]
Indeed, mostly harvesting from websites (I once started working a non-XSLT branch and got somewhere halfway, though I haven't had the time to continue working on it... on hindsight, it would probably have been better to focus on RDFa-Lite first instead of the full version).

E.g. the popular Drupal  CMS has (some basic) support for RDFa (and some attempts for supporting JSON-LD modules, but not in core)

And there is some renewed interest in RDFa from the 'Solid' initiative (https://forum.solidproject.org/t/web-metadata-for-publishing/1729)...

**Comment by knoan** [comment]
Starting some work in this area: will report here if eventually able to submit a PR…

**Comment by edwardsph** [comment]
Just to flag this up - I have had success using https://github.com/semarglproject/semargl-rdf4j as a parser plugin for Rio. The only issue was that it wasn't clear which version of RDFa was being used but when I started getting parsing problems I realised I had to configure it to use RDFa1.1 and it's all good now:
```
  ParserConfig parserConfig = new ParserConfig();
  parserConfig.set(RDFaParserSettings.RDFA_COMPATIBILITY, RDFaVersion.RDFA_1_1);
  Model model = Rio.parse(is, "https://example.org", RDFFormat.RDFA,
                parserConfig, SimpleValueFactory.getInstance(), new ParseErrorLogger());
```
